examples/tokio/main.rs

```diff
@@ -0,0 +1,162 @@
+use anyhow::{anyhow, Error};
+use std::future::Future;
+use tokio::time::Duration;
+use wasmtime::{Config, Engine, Linker, Module, Store};
+// For this example we want to use the async version of wasmtime_wasi.
+// Notably, this version of wasi uses a scheduler that will async yield
+// when sleeping in `poll_oneoff`.
+use wasmtime_wasi::tokio::{Wasi, WasiCtxBuilder};
+
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    // Create an environment shared by all wasm execution. This contains
+    // the `Engine` and the `Module` we are executing.
+    let env = Environment::new()?;
+
+    // The inputs to run_wasm are `Send`: we can create them here and send
+    // them to a new task that we spawn.
+    let inputs1 = Inputs::new(env.clone(), "Gussie");
+    let inputs2 = Inputs::new(env.clone(), "Willa");
+    let inputs3 = Inputs::new(env, "Sparky");
+
+    // Spawn some tasks. Insert sleeps before run_wasm so that the
+    // interleaving is easy to observe.
+    let join1 = tokio::task::spawn(async move { run_wasm(inputs1).await });
+    let join2 = tokio::task::spawn(async move {
+        tokio::time::sleep(Duration::from_millis(750)).await;
+        run_wasm(inputs2).await
+    });
+    let join3 = tokio::task::spawn(async move {
+        tokio::time::sleep(Duration::from_millis(1250)).await;
+        run_wasm(inputs3).await
+    });
+
+    // All tasks should join successfully.
+    join1.await??;
+    join2.await??;
+    join3.await??;
+    Ok(())
+}
+
+#[derive(Clone)]
+struct Environment {
+    engine: Engine,
+    module: Module,
+}
+
+impl Environment {
+    pub fn new() -> Result<Self, Error> {
+        let mut config = Config::new();
+        // We need this engine's `Store`s to be async, and consume fuel, so
+        // that they can co-operatively yield during execution.
+        config.async_support(true);
+        config.consume_fuel(true);
+
+        // Install the host functions for `Wasi`.
+        Wasi::add_to_config(&mut config);
+
+        let engine = Engine::new(&config)?;
+        let module = Module::from_file(&engine, "target/wasm32-wasi/debug/tokio-wasi.wasm")?;
+
+        Ok(Self { engine, module })
+    }
+}
+
+struct Inputs {
+    env: Environment,
+    name: String,
+}
+
+impl Inputs {
+    fn new(env: Environment, name: &str) -> Self {
+        Self {
+            env,
+            name: name.to_owned(),
+        }
+    }
+}
+
+fn run_wasm(inputs: Inputs) -> impl Future<Output = Result<(), Error>> {
```

Comment:
Could this function be shortened to basically "FIXME needs https://github.com/alexcrichton/rfcs-2/blob/new-api/accepted/new-api.md to be safe" or something along those lines?

(probably not too worth maintaining docs here about the unsafety)